pkg/cli/image/manifest/security.go

```diff
@@ -0,0 +1,238 @@
+	flags.StringVar(&o.ICSPFile, "icsp-file", o.ICSPFile, "Path to an ImageContentSourcePolicy file.  If set, data from this file will be used to set alternative image sources. Cannot be set together with --lookup-cluster-icsp=true.")
+}
+
+// ReferentialHTTPClient returns an http.Client that is appropriate for accessing
+// blobs referenced outside of the registry (due to the present of the URLs attribute
+// in the manifest reference for a layer).
+func (o *SecurityOptions) ReferentialHTTPClient() (*http.Client, error) {
+	regContext, err := o.Context()
+	if err != nil {
+		return nil, err
+	}
+	client := &http.Client{}
+	if o.Insecure {
+		client.Transport = regContext.InsecureTransport
+	} else {
+		client.Transport = regContext.Transport
+	}
+	return client, nil
+}
+
+func (o *SecurityOptions) Complete(f kcmdutil.Factory) error {
+	if o.LookupClusterICSP && len(o.ICSPFile) > 0 {
+		return fmt.Errorf("cannot set both --lookup-cluster-icsp=true and --icsp-file")
+	}
+	o.ICSPClientFn = func() (operatorv1alpha1client.ImageContentSourcePolicyInterface, error) {
+		// If ImageContentSourceFile is given, only add ImageContentSource from file, don't search cluster ICSP
+		if len(o.ICSPFile) != 0 {
+			return nil, nil
+		}
+		restConfig, err := f.ToRESTConfig()
+		if err != nil {
+			// may or may not be connected to a cluster
+			// don't error if can't connect
+			klog.V(4).Infof("did not connect to an OpenShift 4.x server, will not lookup ImageContentSourcePolicies: %v", err)
+			return nil, nil
+		}
+		icspClient, err := operatorv1alpha1client.NewForConfig(restConfig)
+		if err != nil {
+			// may or may not be connected to a cluster
+			// don't error if can't connect
+			klog.V(4).Infof("did not connect to an OpenShift 4.x server, will not lookup ImageContentSourcePolicies: %v", err)
+			return nil, nil
+		}
+		return icspClient.ImageContentSourcePolicies(), nil
+	}
+	return nil
+}
+
+func (o *SecurityOptions) Context() (*registryclient.Context, error) {
+	if o.CachedContext != nil {
+		return o.CachedContext, nil
+	}
+	context, err := o.NewContext()
+	if err == nil {
+		o.CachedContext = context
+		o.CachedContext.Retries = 3
+	}
+	return context, err
+}
+
+func (o *SecurityOptions) NewContext() (*registryclient.Context, error) {
+	rt, err := rest.TransportFor(&rest.Config{})
+	if err != nil {
+		return nil, err
+	}
+	insecureRT, err := rest.TransportFor(&rest.Config{TLSClientConfig: rest.TLSClientConfig{Insecure: true}})
+	if err != nil {
+		return nil, err
+	}
+	creds := dockercredentials.NewLocal()
+	if len(o.RegistryConfig) > 0 {
+		creds, err = dockercredentials.NewFromFile(o.RegistryConfig)
+		if err != nil {
+			return nil, fmt.Errorf("unable to load --registry-config: %v", err)
+		}
+	}
+
+	imageSources := registryclient.AlternativeImageSources{
+		ICSPFile:     o.ICSPFile,
+		ICSPList:     o.ICSPList,
+		ICSPClientFn: o.ICSPClientFn,
+	}
+
+	context := registryclient.NewContext(rt, insecureRT, imageSources).WithCredentials(creds)
```

Comment:
I'd imagine we could do `WithICSPFile` and `WithClusterICSP` methods similar to `WithCredentials` above and those would be set based on `--icsp-file` and/or `--lookup-cluster-icsp` flags. And the `Context` method will know both the order (iow. what is more important file vs cluster ICSP) and if those are set and should be used. I think that's what @smarterclayton had in mind and how I understood him.

Reply:
i see, ok, lemme update that

Reply:
updated, and see above comment